src/Utilities/StaticCache.hpp

```diff
@@ -12,119 +12,166 @@
+  using value_type = size_t;
 };
 
-template <typename T>
-class StaticCacheImpl<T> {
- public:
-  template <typename Generator, typename... Args,
-            Requires<not cpp17::is_same_v<std::decay_t<Generator>,
-                                          StaticCacheImpl>> = nullptr>
-  explicit constexpr StaticCacheImpl(Generator&& generator,
-                                     const Args... args) noexcept
-      : data_(generator(args...)) {}
-
-  const T& operator()() const noexcept { return data_; }
-
- private:
-  T data_;
+/// \ingroup UtilitiesGroup
+/// Possible enumeration values for the StaticCache. Only values specified here
+/// are retrievable.
+///
+/// \note The `EnumerationType` must be streamable.
+template <typename EnumerationType, EnumerationType... Enums>
+struct CacheEnumeration {
+  constexpr static size_t size = sizeof...(Enums);
+  using value_type = EnumerationType;
 };
-}  // namespace StaticCache_detail
 
 /// \ingroup UtilitiesGroup
 /// A cache of objects intended to be stored in a static variable.
 ///
-/// Objects can be accessed using several `size_t` arguments in the
-/// ranges specified as template parameters.  The CacheRange template
-/// parameters give first and one-past-last values for the valid
-/// ranges for each argument.
+/// Objects can be accessed via a combination of several `size_t` arguments in
+/// the ranges specified as template parameters of type `CacheRange<start, end>`
+/// and/or enumeration members in the collections of type
+/// `CacheEnumeration<EnumerationType, Members...>`.  The CacheRange template
+/// parameters give first and one-past-last values for the valid ranges for each
+/// argument. The `CacheEnumeration` template parameters are the enumeration
+/// type and an explicit set of every enum member to be cached.
 ///
 /// \example
+/// A cache with only numeric indices:
 /// \snippet Test_StaticCache.cpp static_cache
 ///
+/// \example
+/// A cache with enumeration indices:
+/// \snippet Test_StaticCache.cpp static_cache_with_enum
+///
+/// \example
+/// A cache with mixed numeric and enumeration indices:
+/// \snippet Test_StaticCache.cpp static_cache_with_enum_and_numeric
+///
 /// \see make_static_cache
 ///
 /// \tparam T type held in the cache
 /// \tparam Ranges ranges of valid indices
-template <typename T, typename... Ranges>
+template <typename Generator, typename T, typename... Ranges>
 class StaticCache {
  public:
-  /// Initialize the cache.  All objects will be created by calling
-  /// `generator` before the constructor returns.
-  // clang-tidy: misc-forwarding-reference-overload - fixed with
-  // Requires, but clang-tidy can't recognize that.
-  template <typename Generator,
-            Requires<not cpp17::is_same_v<std::decay_t<Generator>,
-                                          StaticCache>> = nullptr>
-  explicit constexpr StaticCache(Generator&& generator) noexcept  // NOLINT
-      : data_(generator) {}
+  template <typename Gen>
```

Comment:
it allows a slightly different type structure in the passed argument vs the `decay_t<Generator>` passed as a template argument in the function `make_static_cache`. For instance, some of the current uses of the cache wouldn't type match without the additional flexibility because they pass a const lambda. My understanding is that doing it without the decay_t and template arguments, etc, can cause unnecessary copies of the generator.

Reply:
Even worse, without the decay the member variable can be a reference to an object that goes out of scope and so you get a dangling reference :)